File: agent-app/src/main/kotlin/com/riwonace/agent/router/RuleBasedRouter.kt
package com.riwonace.agent.router

import org.springframework.stereotype.Component

enum class Route { SQL, VECTOR, GRAPH }

/**
 * 규칙 기반 도구 라우터 (MCP Parallel 패턴).
 *
 * 소형 LLM의 불안정한 function-calling 대신 결정적 규칙으로 도구를 선택한다.
 * 복수 규칙에 걸리면 해당 도구들을 병렬 호출하고 결과를 통합한다.
 * 어떤 규칙에도 걸리지 않으면 `fallback`이 있으면 그쪽에 위임하고, 없으면 vector_search가 기본값이다.
 */
@Component
class RuleBasedRouter(private val fallback: RouteFallback? = null) {

    private val sqlKeywords = listOf(
        "몇", "개수", "수는", "평균", "합계", "총", "최대", "최소", "가장 비싼", "가장 싼",
        "가장 많은", "가장 높은", "순위", "상위", "하위", "목록", "리스트", "재고", "급여",
        "연봉", "매출", "주문", "계약", "티켓", "우선순위", "분기", "금액", "등록된",
        "직원", "제품", "가격", "부서별", "언제 입사", "입사일",
        "count", "average", "sum", "max", "min", "how many", "list", "top",
    )

    private val graphKeywords = listOf(
        "관계", "연관", "연결", "관련이", "관련된", "의존", "누가 만들", "누가 개발",
        "무엇을 개발", "어디서 만들", "개발사", "라이선스", "무슨 사이",
        "사용 중인", "사용하는", "사용 고객", "이용 고객", "실제 사용", "실제 이용",
        "소속", "담당", "팀장", "이끄", "이슈",
        "relation", "related", "depends", "who made", "who developed",
    )

    private val vectorKeywords = listOf(
        "무엇", "뭐야", "뭔가요", "설명", "소개", "개념", "원리", "어떻게 동작", "왜",
        "장점", "단점", "차이", "정책", "가이드", "방법",
        "장애", "사례", "내용", "취약점", "미팅", "논의", "제안서", "매뉴얼", "인증",
        "설치", "배포", "운영", "백업", "튜닝", "API", "컨테이너", "문서",
        "what is", "explain", "describe", "how does", "why",
    )

    fun route(question: String): List<Route> {
        val q = question.lowercase()
        val routes = buildList {
            if (sqlKeywords.any { q.contains(it) }) add(Route.SQL)
            if (graphKeywords.any { q.contains(it) }) add(Route.GRAPH)
            if (vectorKeywords.any { q.contains(it) }) add(Route.VECTOR)
        }
        if (routes.isEmpty()) return fallback?.classify(question) ?: listOf(Route.VECTOR)

        // 결정 규칙 하나만 잡혔더라도 질문이 여러 요구를 접속하면 의미 분류로 누락 경로만 보강한다.
        // 이미 두 경로 이상이 명확하면 추가 LLM 호출 없이 결정 결과를 유지한다.
        if (routes.size == 1 && fallback != null && COMPOSITION_CUES.any(q::contains)) {
            return (routes + fallback.classify(question)).distinct()
        }
        return routes
    }

    companion object {
        private val COMPOSITION_CUES = listOf("함께", "같이", "동시에", "한 번에", "구분", "그리고", " 및 ", "와 ", "과 ")
    }
}

File: agent-app-go/internal/router/rule_based_router.go
package router

import (
	"regexp"
	"strings"
)

// sqlKeywords mirrors RuleBasedRouter.kt's sqlKeywords list, in the same order
// (order doesn't affect the `any` match, but is kept for side-by-side diffing).
var sqlKeywords = []string{
	"몇", "개수", "수는", "평균", "합계", "총", "최대", "최소", "가장 비싼", "가장 싼",
	"가장 많은", "가장 높은", "순위", "상위", "하위", "목록", "리스트", "재고", "급여",
	"연봉", "매출", "주문", "계약", "티켓", "우선순위", "분기", "금액", "등록된",
	"직원", "가격", "출시 상태", "부서별", "언제 입사", "입사일",
	"count", "average", "sum", "max", "min", "how many", "list", "top",
}

var graphKeywords = []string{
	"관계", "연관", "연결", "관련이", "관련된", "의존", "누가 만들", "누가 개발",
	"무엇을 개발", "어디서 만들", "개발사", "라이선스", "무슨 사이",
	"사용 중인", "사용하는", "사용 고객", "이용 고객", "실제 사용", "실제 이용",
	"이용하는", "사용자 고객",
	"소속", "담당", "팀장", "이끄", "이슈",
	"relation", "related", "depends", "who made", "who developed",
}

var vectorKeywords = []string{
	"무엇", "뭐야", "뭔가요", "설명", "소개", "개념", "원리", "어떻게 동작", "왜",
	"장점", "단점", "차이", "정책", "가이드", "방법",
	"장애", "사례", "내용", "취약점", "미팅", "논의", "제안", "매뉴얼", "인증",
	"설치", "배포", "운영", "보관", "구동", "튜닝", "cpu", "hpa", "컨테이너",
	"what is", "explain", "describe", "how does", "why",
}

var compositionCues = []string{"함께", "같이", "동시에", "한 번에", "구분", "그리고", " 및 ", "와 ", "과 "}

// productNumericCore mirrors PRODUCT_NUMERIC in RuleBasedRouter.kt, minus the
// `(?<![a-z0-9_-])` negative lookbehind that Go's RE2 engine cannot express directly.
// matchesProductNumeric below re-implements that lookbehind by checking the character
// immediately preceding each candidate match.
var productNumericCore = regexp.MustCompile(`(?i)\d[\d,]*(?:\.\d+)?(?:원|만원|%|인|개|건|명)?[^.?!]{0,20}(?:제품|product)`)

var lookbehindExcluded = regexp.MustCompile(`(?i)[a-z0-9_-]`)

// matchesProductNumeric mirrors PRODUCT_NUMERIC.containsMatchIn(q), including the
// `(?<![a-z0-9_-])` guard before the leading digit.
func matchesProductNumeric(q string) bool {
	locs := productNumericCore.FindAllStringIndex(q, -1)
	for _, loc := range locs {
		start := loc[0]
		if start == 0 {
			return true
		}
		prevChar := q[start-1 : start]
		if !lookbehindExcluded.MatchString(prevChar) {
			return true
		}
	}
	return false
}

// RouteFallback mirrors the RouteFallback fun interface: classify a question when no
// deterministic keyword rule matched.
type RouteFallback interface {
	Classify(question string) []Route
}

// RuleBasedRouter ports RuleBasedRouter.kt 1:1: deterministic keyword/regex routing,
// with an optional fallback for the no-match and single-match-composition cases.
type RuleBasedRouter struct {
	Fallback RouteFallback
}

func containsAny(q string, keywords []string) bool {
	for _, k := range keywords {
		if strings.Contains(q, k) {
			return true
		}
	}
	return false
}

// Route mirrors fun route(question: String): List<Route>.
func (r *RuleBasedRouter) Route(question string) []Route {
	q := strings.ToLower(question)

	var routes []Route
	if containsAny(q, sqlKeywords) || matchesProductNumeric(q) {
		routes = append(routes, RouteSQL)
	}
	if containsAny(q, graphKeywords) {
		routes = append(routes, RouteGraph)
	}
	if containsAny(q, vectorKeywords) {
		routes = append(routes, RouteVector)
	}

	if len(routes) == 0 {
		if r.Fallback != nil {
			return r.Fallback.Classify(question)
		}
		return []Route{RouteVector}
	}

	if len(routes) == 1 && r.Fallback != nil && containsAny(q, compositionCues) {
		return dedupRoutes(append(routes, r.Fallback.Classify(question)...))
	}

	return routes
}

func dedupRoutes(routes []Route) []Route {
	seen := make(map[Route]bool)
	var out []Route
	for _, r := range routes {
		if !seen[r] {
			seen[r] = true
			out = append(out, r)
		}
	}
	return out
}
